### PR TITLE
Use GPUtil for gpu detection when available

### PR DESCRIPTION
**Pull request by mopga**
## Why are these changes needed?

In Envs with K8S and enabled SELinux there is a bug:
"/proc/nvidia/" is not allowed to mount in container
So, i made a rework for GPU detection based on GPutil package.

## Checks

- [x] I've run `scripts/format.sh` to lint the changes in this PR.
- [x] I've made sure the tests are passing. Note that there might be a few flaky tests, see the recent failures at https://flakey-tests.ray.io/
- Testing Strategy
   - [x] Release tests

### CODE REVIEW

**Comment by richardliaw** [comment]
@mopga can you provide some more context about what this problem solves?

Like, why is SELinux problematic, is it a very common issue?

**Comment by mopga** [comment]
@richardliaw  our case is:
We deploy our JupyterLab to K8S cluster
in deployment policy we attached our GPUs (NVIDIA TESLA V100/P100) to container with JupyterLab
When is strats we installed ray by
`pip install ray`
then we try to start ray, for example:
`ray start --head`
or in code
`ray.init()`
and it throws an error:

Traceback (most recent call last):
  File "/opt/app-root/bin/ray", line 8, in <module>
    sys.exit(main())
  File "/opt/app-root/lib/python3.6/site-packages/ray/scripts/scripts.py", line 1981, in main
    return cli()
  File "/opt/app-root/lib/python3.6/site-packages/click/core.py", line 1137, in __call__
    return self.main(*args, **kwargs)
  File "/opt/app-root/lib/python3.6/site-packages/click/core.py", line 1062, in main
    rv = self.invoke(ctx)
  File "/opt/app-root/lib/python3.6/site-packages/click/core.py", line 1668, in invoke
    return _process_result(sub_ctx.command.invoke(sub_ctx))
  File "/opt/app-root/lib/python3.6/site-packages/click/core.py", line 1404, in invoke
    return ctx.invoke(self.callback, **ctx.params)
  File "/opt/app-root/lib/python3.6/site-packages/click/core.py", line 763, in invoke
    return __callback(*args, **kwargs)
  File "/opt/app-root/lib/python3.6/site-packages/ray/scripts/scripts.py", line 641, in start
    ray_params, head=True, shutdown_at_exit=block, spawn_reaper=block)
  File "/opt/app-root/lib/python3.6/site-packages/ray/node.py", line 248, in __init__
    self.start_head_processes()
  File "/opt/app-root/lib/python3.6/site-packages/ray/node.py", line 894, in start_head_processes
    self.start_redis()
  File "/opt/app-root/lib/python3.6/site-packages/ray/node.py", line 705, in start_redis
    self.get_resource_spec(),
  File "/opt/app-root/lib/python3.6/site-packages/ray/node.py", line 374, in get_resource_spec
    is_head=self.head, node_ip_address=self.node_ip_address)
  File "/opt/app-root/lib/python3.6/site-packages/ray/resource_spec.py", line 149, in resolve
    num_gpus = _autodetect_num_gpus()
  File "/opt/app-root/lib/python3.6/site-packages/ray/resource_spec.py", line 237, in _autodetect_num_gpus
    result = len(os.listdir(proc_gpus_path))
PermissionError: [Errno 13] Permission denied: '/proc/driver/nvidia/gpus'

**Review by wuisawesome — CHANGES_REQUESTED** [review]
Having more robust gpu detection logic seems like it would be good, but I'm a bit worried about introducing a new ray core dependency. can we make it optional by only going through this new logic if we detect the `GPUtil` exists?

```
if importlib.util.find_spec("GPUtil"):
   ...
```

@richardliaw maybe you can recommend which extras we would include `GPUtil` in.

**Comment by heinerb** [comment]
Update on this item? Is this looking to be merged?

**Comment by mopga** [comment]
> Update on this item? Is this looking to be merged?
i'll try to fix dependencies in proper place. sorry, lot of work all the time.

**Comment by heinerb** [comment]
@mopga the fix in this issue does appear to resolve issue in https://github.com/ray-project/ray/issues/17914

**Comment by wuisawesome** [comment]
@heinerb bump on https://github.com/ray-project/ray/pull/18938#pullrequestreview-782341617

**Comment by bveeramani** [comment]
## :bangbang: ACTION REQUIRED :bangbang:

We've switched our code formatter from YAPF to Black (see #21311).

To prevent issues with merging your code, here's what you'll need to do:

1. Install Black
```
pip install -I black==21.12b0
```

2. Format changed files with Black
```
curl -o format-changed.sh https://gist.githubusercontent.com/bveeramani/42ef0e9e387b755a8a735b084af976f2/raw/7631276790765d555c423b8db2b679fd957b984a/format-changed.sh
chmod +x ./format-changed.sh
./format-changed.sh
rm format-changed.sh
```

3. Commit your changes.
```
git add --all
git commit -m "Format Python code with Black"
```

4. Merge master into your branch.
```
git pull upstream master
```

5. Resolve merge conflicts (if necessary).

After running these steps, you'll have the updated `format.sh`.

File: python/requirements/ml/requirements_tune.txt [hunk 1]
```diff
@@ -40,3 +40,8 @@ transformers==4.10.0
 wandb==0.12.5
 xgboost==1.3.3
 zoopt==0.4.1
+# Dependencies for Hugging Face examples:
```
Comment (wuisawesome) [inline python/requirements/ml/requirements_tune.txt:43]: @richardliaw ptal

**Review by wuisawesome — APPROVED** [review]
Thanks for cleaning this up! Approving since I'm happy with it, but will wait for @richardliaw to take a look at the requirements changes for tune.

File: python/requirements/ml/requirements_tune.txt (hunk 1 above)
Comment (juliusfrost) [reply python/requirements/ml/requirements_tune.txt]: This was a result of a faulty merge and shouldn't be added. See https://github.com/ray-project/ray/pull/20767

**Comment by wuisawesome** [comment]
@mopga looks like the change is causing some test failures:
```


=================================== FAILURES ===================================
--
  | _________________________________ test_output __________________________________
  |  
  | def test_output():
  | # Use subprocess to execute the __main__ below.
  | outputs = subprocess.check_output(
  | [sys.executable, __file__, "_ray_instance"], stderr=subprocess.STDOUT
  | ).decode()
  | lines = outputs.split("\n")
  | for line in lines:
  | print(line)
  | if os.environ.get("RAY_MINIMAL") == "1":
  | # Without "View the Ray dashboard"
  | assert len(lines) == 1, lines
  | else:
  | # With "View the Ray dashboard"
  | >           assert len(lines) == 2, lines
  | E           AssertionError: ['2022-02-22 22:57:19,504\tINFO services.py:1496 -- View the Ray dashboard at \x1b[1m\x1b[32mhttp://127.0.0.1:8265\x1b...514-1/ray-project/ray-builders-pr/python/ray/_private/resource_spec.py", line 302, in _constraints_from_gpu_info', ...]
  | E           assert 9 == 2
  | E             +9
  | E             -2
  |  
  | /Users/ec2-user/.buildkite-agent/builds/bk-mac1-experimental-queue-i-0313233ecf4829514-1/ray-project/ray-builders-pr/python/ray/tests/test_output.py:365: AssertionError
  | ----------------------------- Captured stdout call -----------------------------
  | 2022-02-22 22:57:19,504	INFO services.py:1496 -- View the Ray dashboard at http://127.0.0.1:8265
  | 2022-02-22 22:57:19,506	ERROR resource_spec.py:188 -- Could not parse gpu information.
  | Traceback (most recent call last):
  | File "/Users/ec2-user/.buildkite-agent/builds/bk-mac1-experimental-queue-i-0313233ecf4829514-1/ray-project/ray-builders-pr/python/ray/_private/resource_spec.py", line 185, in resolve
  | gpu_types = _constraints_from_gpu_info(info_string)
  | File "/Users/ec2-user/.buildkite-agent/builds/bk-mac1-experimental-queue-i-0313233ecf4829514-1/ray-project/ray-builders-pr/python/ray/_private/resource_spec.py", line 302, in _constraints_from_gpu_info
  | if len(info_str) <= 0:
  | TypeError: object of type 'NoneType' has no len()
  |  
  | =============================== warnings summary ===============================
  | /usr/local/opt/miniconda/lib/python3.6/site-packages/_pytest/config/__init__.py:935
  | /usr/local/opt/miniconda/lib/python3.6/site-packages/_pytest/config/__init__.py:935: PytestAssertRewriteWarning: Module already imported so cannot be rewritten: anyio
  | self._mark_plugins_for_rewrite(hook)
  |  
  | -- Docs: https://docs.pytest.org/en/latest/warnings.html
  | =========================== short test summary info ============================
  | FAILED ::test_output - AssertionError: ['2022-02-22 22:57:19,504\tINFO servic...


```

**Comment by juliusfrost** [comment]
I've fixed that test @wuisawesome

**Comment by wuisawesome** [comment]
Thanks! tests look good. the minimal install failure looks like the grpc pinning issue that was broken in master. Rllib failures look like the pendulum tests that are all currently broken.

**Pull request merged**